Use WeakRefs for spawned tasks to avoid holding unexpected references

quinnj opened the pull request:
Most complete explanation is [here](https://github.com/JuliaLang/julia/issues/40626#issuecomment-1340376861).

Also discussed [here](https://github.com/JuliaData/CSV.jl/issues/1057).

This PR proposes an alternative solution to `clear_thread_states` where that approach can be problematic (interferring with global thread state, not working as expected in nested spawned tasks, etc.). The previous definition also started unending `Timer` tasks that could build up over time.

The approach in this PR is to wrap spawned task closure arguments in `WeakRef` to allow them to be GCed as expected once the tasks are finished.

bkamins commented on src/utils.jl:
````diff
@@ -616,23 +616,45 @@ macro refargs(ex)
+but with an `_` prefix; this avoids the original input being capture by `Task` closure. So usage is like:
+
+```julia
+@weakrefspawn ctx foo begin
````
maybe add the `@sync` part around it to give a full example of intended use?

bkamins commented on src/utils.jl:
```diff
@@ -616,23 +616,45 @@ macro refargs(ex)
+    end
+
+A macro that wraps a `Threads.@spawn` block with `WeakRef` calls for all `args...` arguments, allowing
+them to be garbage collected once the `Task` has finished running. Must be used within a `@syncpreserve`
```
Thank you for working on this!

Above `@sync` not `@syncpreserve` is used. Which is correct?

Also are we sure that the "preserve" part touches only the selected variables (this is relevant, as inside the task GC might have to be run if the taks itself allocates a lot).

quinnj commented:
@bkamins, I've updated this PR to use the `@wkspawn` macro now provided in the WorkerUtilities.jl package: https://github.com/JuliaServices/WorkerUtilities.jl/pull/10. I moved it there because I have a few other multithreaded/concurrency-related utilities there and I want to use this in the Arrow.jl package as well. 

As noted in that PR, the `@wkspawn` macro is a drop-in replacement for `Threads.@spawn`, but will pass mutable closure arguments wrapped in `WeakRef` to the spawn block expression.
